File: java/p2/src/p2/core/Driver.java
package p2.core;

import java.util.ArrayList;
import java.util.Hashtable;
import java.util.List;
import java.util.Set;
import p2.exec.Query;
import p2.lang.plan.Program;
import p2.lang.plan.Watch.WatchTable;
import p2.types.basic.Tuple;
import p2.types.basic.TupleSet;
import p2.types.basic.TypeList;
import p2.types.exception.P2RuntimeException;
import p2.types.exception.UpdateException;
import p2.types.operator.Operator;
import p2.types.operator.Watch;
import p2.types.table.Aggregation;
import p2.types.table.Table;
import p2.types.table.TableName;
import p2.lang.Compiler;

public class Driver implements Runnable {
	
	public static class Flusher extends p2.types.table.Function {
		private static class ScheduleUnit {
			public Long time;
			public String program;
			public TableName name;
			public TupleSet insertions;
			public TupleSet deletions;
			
			public ScheduleUnit(Tuple tuple) {
				this.time    = (Long)      tuple.value(Field.TIME.ordinal());
				this.program = (String)    tuple.value(Field.PROGRAM.ordinal());
				this.name    = (TableName) tuple.value(Field.TABLENAME.ordinal());
				insertions = new TupleSet(name);
				deletions = new TupleSet(name);
			}
			
			public void add(Tuple tuple) {
				TupleSet  insertions = (TupleSet)  tuple.value(Field.INSERTIONS.ordinal());
				TupleSet  deletions  = (TupleSet)  tuple.value(Field.DELETIONS.ordinal());
				if (insertions != null) this.insertions.addAll(insertions);
				if (deletions != null) this.deletions.addAll(deletions);
			}
			
			public Tuple tuple() {
				return new Tuple(this.time, this.program, this.name, this.insertions, this.deletions);
			}
			
			@Override
			public int hashCode() {
				return toString().hashCode();
			}
			@Override
			public boolean equals(Object o) {
				return o instanceof ScheduleUnit &&
				       toString().equals(o.toString());
			}
			@Override
			public String toString() {
				return this.program + ":" + ":" + time.toString() + ":" +  name;
			}
		}
		
		public TupleSet aggregate(TupleSet tuples) {
			Hashtable<ScheduleUnit, ScheduleUnit> units = new Hashtable<ScheduleUnit, ScheduleUnit>();
			for (Tuple tuple : tuples) {
				ScheduleUnit unit = new ScheduleUnit(tuple);
				if (!units.containsKey(unit)) {
					units.put(unit, unit);
				}
				units.get(unit).add(tuple);
			}
			TupleSet aggregate = new TupleSet(name());
			for (ScheduleUnit unit : units.keySet()) {
				if (unit.insertions.size() > 0 || unit.deletions.size() > 0)
					aggregate.add(unit.tuple());
			}
			return aggregate;
		}
		
		public enum Field{TIME, PROGRAM, TABLENAME, INSERTIONS, DELETIONS};
		public static final Class[] SCHEMA =  {
			Long.class,       // Time
			String.class,     // Program name
			TableName.class,  // Table name
			TupleSet.class,   // Insertion tuple set
			TupleSet.class    // Deletions tuple set
		};
		
		private Runtime context; 

		public Flusher(Runtime context) {
			super("flusher", new TypeList(SCHEMA));
			this.context = context;
		}
		
		public TupleSet insert(TupleSet tuples, TupleSet conflicts) throws UpdateException {
			TupleSet delta = new TupleSet(name());
			for (Tuple tuple : aggregate(tuples)) {
				Long      time       = (Long)      tuple.value(Field.TIME.ordinal());
				String    program    = (String)    tuple.value(Field.PROGRAM.ordinal());
				TableName name       = (TableName) tuple.value(Field.TABLENAME.ordinal());
				TupleSet  insertions = (TupleSet)  tuple.value(Field.INSERTIONS.ordinal());
				TupleSet  deletions  = (TupleSet)  tuple.value(Field.DELETIONS.ordinal());
				
				if (insertions == null) insertions = new TupleSet(name);
				if (deletions == null)  deletions = new TupleSet(name);
				
				if (insertions.size() == 0 && deletions.size() == 0) {
					continue;
				}
				
				WatchTable watch = (WatchTable) context.catalog().table(WatchTable.TABLENAME);
				Table table = context.catalog().table(name);
				if (insertions.size() > 0 || table instanceof Aggregation) {
					insertions = table.insert(insertions, deletions);
					
					if (table instanceof Aggregation) {
						Operator watchRemove = watch.watched(program, name, Watch.Modifier.ERASE);
						if (watchRemove != null) {
							try { watchRemove.evaluate(deletions);
							} catch (P2RuntimeException e) { }
						}
					}
				}
				else { 
					if (table.type() != Table.Type.TABLE) return new TupleSet(name);
					
					deletions = table.delete(deletions);
					
					Operator watchRemove = watch.watched(program, name, Watch.Modifier.ERASE);
					if (watchRemove != null) {
						try { watchRemove.evaluate(deletions);
						} catch (P2RuntimeException e) { }
					}
				}
				
				if (insertions.size() > 0) {
					Operator watchAdd = watch.watched(program, name, Watch.Modifier.ADD);
					if (watchAdd != null) {
						try { watchAdd.evaluate(insertions);
						} catch (P2RuntimeException e) { }
					}
				}
				
				tuple.value(Field.INSERTIONS.ordinal(), insertions);
				tuple.value(Field.DELETIONS.ordinal(), deletions);
				delta.add(tuple);
			}
			return delta;
		}
	}
	
	public static class Evaluator extends p2.types.table.Function {
		public enum Field{TIME, PROGRAM, TABLENAME, INSERTIONS, DELETIONS};
		public static final Class[] SCHEMA =  {
			Long.class,       // Evaluation time
			String.class,     // Program name
			TableName.class,  // Table name
			TupleSet.class,   // Insertion tuple set
			TupleSet.class    // Deletions tuple set
		};
		
		private Runtime context;

		public Evaluator(Runtime context) {
			super("evaluator", new TypeList(SCHEMA));
			this.context = context;
		}
		
		public TupleSet insert(TupleSet tuples, TupleSet conflicts) throws UpdateException {
			TupleSet delta = new TupleSet(name());
			for (Tuple tuple : tuples) {
				Long      time       = (Long)      tuple.value(Field.TIME.ordinal());
				String    programName= (String)    tuple.value(Field.PROGRAM.ordinal());
				TableName name       = (TableName) tuple.value(Field.TABLENAME.ordinal());
				TupleSet  insertions = (TupleSet)  tuple.value(Field.INSERTIONS.ordinal());
				TupleSet  deletions  = (TupleSet)  tuple.value(Field.DELETIONS.ordinal());
				if (deletions == null) deletions = new TupleSet(name);
				Program program = context.program(programName);
				
				if (program != null) {
					TupleSet  result = evaluate(time, program, name, insertions, deletions);
					delta.addAll(result);
				}
				else {
					java.lang.System.err.println("EVALUATOR ERROR: unknown program " + programName + "!");
				}
			}
			return delta;
		}
		
		private TupleSet evaluate(Long time, Program program, TableName name, TupleSet insertions, TupleSet deletions) 
		throws UpdateException {
			Hashtable<String, Tuple> continuations = new Hashtable<String, Tuple>();

			WatchTable watch = (WatchTable) context.catalog().table(WatchTable.TABLENAME);
			Operator watchInsert = watch.watched(program.name(), name, Watch.Modifier.INSERT);
			Operator watchDelete = watch.watched(program.name(), name, Watch.Modifier.DELETE);

			Set<Query> querySet = program.queries(name);
			if (querySet == null) {
				return new TupleSet(name); // Done
			}
			
			if (insertions.size() > 0) {
				/* We're not going to deal with the deletions yet. */
				continuation(continuations, time, program.name(), Table.Event.DELETE, deletions);

				for (Query query : querySet) {
					if (query.event() != Table.Event.DELETE) {
						if (watchInsert != null) {
							try { watchInsert.rule(query.rule()); watchInsert.evaluate(insertions);
							} catch (P2RuntimeException e) { 
								java.lang.System.err.println("WATCH INSERTION FAILURE ON " + name + "!");
							}
						}
						
						TupleSet result = null;
						try {
							result = query.evaluate(insertions);
							if (result.size() == 0) continue;
						} catch (P2RuntimeException e) {
							e.printStackTrace();
							java.lang.System.exit(0);
						}

						if (query.isDelete()) {
							continuation(continuations, time, program.name(), Table.Event.DELETE, result);
						}
						else {
							continuation(continuations, time, program.name(), Table.Event.INSERT, result);
						}
					}
				}
			}
			else if (deletions.size() > 0) {
				for (Query query : querySet) {
					Table output = context.catalog().table(query.output().name());
					if (query.event() == Table.Event.DELETE ||
							(output.type() == Table.Type.TABLE && query.event() != Table.Event.INSERT)) {
						if (watchDelete != null) {
							try { watchDelete.rule(query.rule()); watchDelete.evaluate(deletions);
							} catch (P2RuntimeException e) { }
						}
						
						TupleSet result = null;
						try {
							result = query.evaluate(deletions);
							if (result.size() == 0) continue;
						} catch (P2RuntimeException e) {
							e.printStackTrace();
							java.lang.System.exit(0);
						}
						
						if (!query.isDelete() && output.type() == Table.Type.EVENT) {
							/* Query is not a delete and it's output type is an event. */
							continuation(continuations, time, program.name(), Table.Event.INSERT, result);
						}
						else if (output.type() == Table.Type.TABLE) {
							continuation(continuations, time, program.name(), Table.Event.DELETE, result);
						}
						else {
							throw new UpdateException("Query " + query + " is trying to delete from table " + output.name() + "?");
						}
					}
				}
			}

			TupleSet delta = new TupleSet(name);
			for (Tuple continuation : continuations.values()) {
				TupleSet ins  = (TupleSet) continuation.value(Field.INSERTIONS.ordinal());
				TupleSet dels = (TupleSet) continuation.value(Field.DELETIONS.ordinal());
				if (ins.size() > 0 || dels.size() > 0) {
					delta.add(continuation);
				}
			}

			return delta;
		}

		private void continuation(Hashtable<String, Tuple> continuations, Long time,
				                  String program, Table.Event event, TupleSet result) {
			String key = program + "." + result.name();

			if (!continuations.containsKey(key)) {
				Tuple tuple = new Tuple(time, program, result.name(),
						                new TupleSet(result.name()), 
						                new TupleSet(result.name()));
				continuations.put(key, tuple);
			}

			if (event == Table.Event.INSERT) {
				TupleSet insertions = (TupleSet) continuations.get(key).value(Field.INSERTIONS.ordinal());
				insertions.addAll(result);
			}
			else {
				TupleSet deletions = (TupleSet) continuations.get(key).value(Field.DELETIONS.ordinal());
				deletions.addAll(result);
			}
		}
	}

	public interface Task {
		public TupleSet insertions();
		
		public TupleSet deletions();
		
		public String program();

		public TableName name();
	}
	
	/* Tasks that the driver needs to execute during the next clock. */
	private List<Task> tasks;
	
	/** The schedule queue. */
	private Program runtime;

	private Schedule schedule;

	private Clock clock;
	
	public Evaluator evaluator;
	
	private Flusher flusher;

	public Driver(Runtime context, Schedule schedule, Clock clock) {
		this.tasks = new ArrayList<Task>();
		this.schedule = schedule;
		this.clock = clock;
		this.evaluator = new Evaluator(context);
		this.flusher = new Flusher(context);
		
		context.catalog().register(this.evaluator);
		context.catalog().register(this.flusher);
	}
	
	public void runtime(Program runtime) {
		this.runtime = runtime;
	}

	public void task(Task task) {
		this.tasks.add(task);
	}

	public void run() {
		TupleSet time = clock.time(0L);
		while (true) {
			synchronized (this) {
				try {
					java.lang.System.err.println("============================     EVALUATE SCHEDULE     =============================");
					evaluate(runtime.name(), time.name(), time, null); // Clock insert current
					
					/* Evaluate task queue. */
					for (Task task : tasks) {
						evaluate(task.program(), task.name(), task.insertions(), task.deletions());
					}
					tasks.clear(); // Clear task queue.
					evaluate(runtime.name(), time.name(), null, time); // Clock delete current
					java.lang.System.err.println("============================ ========================== =============================");
				} catch (UpdateException e) {
					e.printStackTrace();
					java.lang.System.exit(1);
				}

				/* Check for new tasks or schedules, if none wait. */
				while (this.tasks.size() == 0 && schedule.cardinality() == 0) {
					try {
						this.wait();
					} catch (InterruptedException e) { }
				}
				if (schedule.cardinality() > 0) {
					time = clock.time(schedule.min());
				}
				else {
					time = clock.time(clock.current() + 1);
				}
			}
		}
	}
	
	public void evaluate(String program, TableName name, TupleSet insertions, TupleSet deletions) throws UpdateException {
		TupleSet insert = new TupleSet();
		TupleSet delete = new TupleSet();
		insert.add(new Tuple(clock.current(), program, name, insertions, deletions)); 
		/* Evaluate until nothing remains. */
		while (insert.size() > 0 || delete.size() > 0) {
			TupleSet delta = null;
			while(insert.size() > 0) {
				delta = flusher.insert(insert, null);
				delta = evaluator.insert(delta, null);
				insert.clear();
				split(delta, insert, delete);
			}
			
			while(delete.size() > 0) {
				delta = flusher.insert(delete, null);
				delta = evaluator.insert(delta, null);
				delete.clear();
				split(delta, insert, delete);
			}
		}
	}
	
	private void split(TupleSet tuples, TupleSet insertions, TupleSet deletions) {
		for (Tuple tuple : tuples) {
			Tuple insert = tuple.clone();
			Tuple delete = tuple.clone();
			insert.value(Evaluator.Field.INSERTIONS.ordinal(), tuple.value(Evaluator.Field.INSERTIONS.ordinal()));
			insert.value(Evaluator.Field.DELETIONS.ordinal(), null);
			delete.value(Evaluator.Field.INSERTIONS.ordinal(), null);
			delete.value(Evaluator.Field.DELETIONS.ordinal(), tuple.value(Evaluator.Field.DELETIONS.ordinal()));
			
			insertions.add(insert);
			deletions.add(delete);
		}
	}


}

File: lincoln/ruby/lib/core/driver.rb
require 'lib/types/table/table_function'
require 'lib/types/basic/tuple_set'
require 'lib/types/table/aggregation_table'
require 'lib/types/table/event_table'
require 'lib/types/operator/watch_op'
require 'monitor'

class Driver < Monitor
  Infinity = 1.0/0
  @@threads = []

  class Flusher < TableFunction
    class ScheduleUnit 
      def initialize(tuple)
        @time    = tuple.value(Field::TIME)
        @program = tuple.value(Field::PROGRAM)
        @name    = tuple.value(Field::TABLENAME)
        @insertions = TupleSet.new(@name)
        @deletions = TupleSet.new(@name)
      end

      attr_accessor :insertions, :deletions
      def <<(tuple)
        insertions = tuple.value(Field::INSERTIONS)
        deletions  = tuple.value(Field::DELETIONS)
        @insertions.addAll(insertions) unless insertions.nil? 
        @deletions.addAll(deletions) unless deletions.nil? 
      end

      def tuple
        Tuple.new(@time, @program, @name, @insertions, @deletions)
      end

      def hash
        to_s.hash
      end

      def ==(o)
        return ((o.class <= ScheduleUnit) and (to_s == o.to_s))
      end

      def to_s
        @program + ":" + ":" + @time.to_s + ":" +  @name.to_s
      end
    end

    #  Aggregates the set of tuples into groups
    #  of tuples that are flushed into the same table.
    #  tuples is the set of tuples to be flushed
    #  returns an Aggregated set of tuples.
    def aggregate(tuples)
      units = Hash.new
      tuples.each do |tuple|
        unit = ScheduleUnit.new(tuple)
        units[unit] = unit unless units.has_key?(unit)
        units[unit] << tuple
      end

      aggregate = TupleSet.new(name)
      units.keys.each do |unit|
        if (unit.insertions.size > 0 || unit.deletions.size > 0)
          aggregate << unit.tuple
        end
      end
      return aggregate
    end

    class Field
      TIME = 0
      PROGRAM = 1
      TABLENAME = 2
      INSERTIONS = 3
      DELETIONS = 4
    end

    @@SCHEMA =  [Integer, String, TableName, TupleSet, TupleSet]

    #  Creates a new flusher table function.
    #  context: The runtime context. 
    def initialize(context)
      super("flusher", TypeList.new(@@SCHEMA))
      @context = context
    end

    def insert(tuples, conflicts)
      delta = TupleSet.new(name)
      aggregate(tuples).each do |tuple| 
        time       = tuple.value(Field::TIME)
        program    = tuple.value(Field::PROGRAM)
        name       = tuple.value(Field::TABLENAME)
        insertions = tuple.value(Field::INSERTIONS)
        deletions  = tuple.value(Field::DELETIONS)

        insertions = TupleSet.new(name) if insertions.nil?
        deletions  = TupleSet.new(name) if deletions.nil?

        next if (insertions.size == 0 and deletions.size == 0)

        watch = @context.catalog.table(WatchTable.table_name)
        table = @context.catalog.table(name)
        if (insertions.size > 0 || table.class <= AggregationTable)
          insertions = table.insert(insertions, deletions)

          if (table.class <= AggregationTable)
            watchRemove = watch.watched(program, name, WatchOp::Modifier::ERASE)
            watchRemove.evaluate(deletions) unless (watchRemove.nil?) 
          end
        else 
          return TupleSet.new(name) if (table.table_type != Table::TableType::TABLE)

          deletions = table.delete(deletions)

          watchRemove = watch.watched(program, name, WatchOp::Modifier::ERASE)
          watchRemove.evaluate(deletions) unless watchRemove.nil?

        end

        if (insertions.size > 0) 
          watchAdd = watch.watched(program, name, WatchOp::Modifier::ADD)
          watchAdd.evaluate(insertions) unless watchAdd.nil?
        end

        tuple.set_value(Field::INSERTIONS, insertions)
        tuple.set_value(Field::DELETIONS, deletions)
        delta << tuple
      end
      return delta
    end
  end

  #  Table function represents the query processor that evaluates 
  #  tuples using the query objects install by the runtime programs.
  #  The tuple format that this function expects is as follows:
  #  <Time, Program, TableName, Insertions, Deletions>
  #  Time: The evaluation time (usually taken from the system clock).
  #  Program: The program whose query objects are to evaluate the tuples.
  #  TableName: The name of the table to which the tuples refer.
  #  Insertions: Contains a set of tuples that represent the delta set
  #  of an insertion into the respective table.
  #  Deletions: Contains a set of tuples representing the delta set
  #  of a deletion from the respective table.
  #  NOTE: Deletions are not evaluated unless there are no insertions
  #  The return value of this table function contains a set of tuples
  #  that represent the output of the query evaluations. 

  class Evaluator < TableFunction
    class Field
      TIME = 0
      PROGRAM = 1
      TABLENAME = 2
      INSERTIONS = 3
      DELETIONS = 4
    end
    @@SCHEMA =  [Integer, String, TableName, TupleSet, TupleSet]

    #  Creates a new evaluated based on the given runtime.
    #  @param context The runtime context. 
    def initialize(context)
      super("evaluator", TypeList.new(@@SCHEMA))
      @context = context
    end

    def insert(tuples, conflicts)
      delta = TupleSet.new(name)
      tuples.each do |tuple|
        time       = tuple.value(Field::TIME)
        programName    = tuple.value(Field::PROGRAM)
        name       = tuple.value(Field::TABLENAME)
        insertions = tuple.value(Field::INSERTIONS)
        deletions  = tuple.value(Field::DELETIONS)

        deletions  = TupleSet.new(name) if deletions.nil?

        program = @context.program(programName)

        if !program.nil?
          result = evaluate(time, program, name, insertions, deletions)
          delta.addAll(result)
        else
          raise("EVALUATOR ERROR: unknown program " + programName + "!")
        end
      end
      return delta
    end

    #  Evaluates the given insertions/deletions against the program queries.
    #  @param time The current time.
    #  @param program The program whose queries will evaluate the tuples.
    #  @param name The name of the table to which the tuples refer.
    #  @param insertions A delta set of tuple insertions.
    #  @param deletions A delta set of tuple deletions.
    #  @return The result of the query evaluation. NOTE: if insertions.size > 0 then
    #  deletions will not be evaluated but rather deferred until a call is made with
    #  insertions.size == 0.
    #  @throws UpdateException On evaluation error.
    def evaluate(time, program, name, insertions, deletions) 
      continuations = Hash.new

      watch = @context.catalog.table(WatchTable.table_name)
      watchInsert = watch.watched(program.name, name, WatchOp::Modifier::INSERT)
      watchDelete = watch.watched(program.name, name, WatchOp::Modifier::DELETE)

      querySet = program.get_queries(name)
      if (querySet.nil?)
        return TupleSet.new(name) # Done
      end

      if (insertions.size > 0) 
        # We're not going to deal with the deletions yet.
        continuation(continuations, time, program.name, Predicate::Event::DELETE, deletions)

        querySet.each do |query|
          if (query.event != Predicate::Event::DELETE)
            if (!watchInsert.nil?)
              watchInsert.rule(query.rule)
              watchInsert.evaluate(insertions)
            end
            result = query.evaluate(insertions)
            next if (result.size == 0)

            if (query.isDelete)
              continuation(continuations, time, program.name, Predicate::Event::DELETE, result)
            else 
              continuation(continuations, time, program.name, Predicate::Event::INSERT, result)
            end
          end
        end
      elsif (deletions.size > 0)
        querySet.each do |query|
          output = @context.catalog.table(query.output.name)
          if (query.event == Predicate::Event::DELETE or
            (output.table_type == Table::TableType::TABLE and query.event != Predicate::Event::INSERT))
            if (!watchDelete.nil?) 
              watchDelete.rule(query.rule)
              watchDelete.evaluate(deletions)
            end
            result = query.evaluate(deletions)
            next if (result.size == 0)

            if (!query.isDelete && output.table_type == Table::TableType::EVENT) 
              # Query is not a delete and it's output type is an event.
              continuation(continuations, time, program.name, Predicate::Event::INSERT, result)
            elsif (output.table_type == Table::TableType::TABLE)
              continuation(continuations, time, program.name, Predicate::Event::DELETE, result)
            else
              raise "Query " + query + " is trying to delete from table " + output.name + "?"
            end
          end
        end
      end

      delta = TupleSet.new(name)
      continuations.values.each do |continuation|
        ins  = continuation.value(Field::INSERTIONS)
        dels = continuation.value(Field::DELETIONS)
        delta << continuation if (ins.size > 0 || dels.size > 0)
      end

      return delta
    end

    # Helper routine that packages up result tuples grouped by (time, program, tablename).
    # @param continuations Hashtable containing the tuple groups
    # @param time Current time.
    # @param program Program that evaluated the tuples.
    # @param event Indicates whether the tuples are insertions or deletions.
    # @param result The result tuples taken from the output of a query.
    def continuation(continuations, time, program, event, result)
      key = program.to_s + "." + result.name.to_s

      if (!continuations.has_key?(key))
        tuple = Tuple.new(time, program, result.name, TupleSet.new(result.name), TupleSet.new(result.name))
        continuations[key] = tuple
      end

      if (event == Predicate::Event::INSERT)
        insertions = continuations[key].value(Field::INSERTIONS)
        insertions.addAll(result)
      else
        deletions = continuations[key].value(Field::DELETIONS)
        deletions.addAll(result)
      end
    end

  end
   # Tasks are used to inject tuples into the schedule.
  class Task 
    attr_reader :insertions, :deletions, :program, :name
  end

  # Creates a new driver.
  # @param context The runtime context.
  # @param schedule The schedule table.
  # @param clock The system clock table.
  def initialize(context, schedule, clock)
    @tasks = Array.new
    @schedule = schedule
    @clock = clock
    @evaluator = Evaluator.new(context)
    @flusher = Flusher.new(context)

    context.catalog.register(@evaluator)
    context.catalog.register(@flusher)
    # important to call super on Monitor initialization
    super()
  end

  # Set the runtime program.
  # @param runtime The runtime program. 
  attr_accessor :runtime

  # * Add a task to the task queue. This will be evaluated
  # * on the next clock tick.
  # * @param task The task to be added.
  def task(task)
    @tasks << task
  end

  # * The main driver loop.
  # * The loop is responsible for 1. updating the clock 2. scheduling tasks.
  # * Driver algorithm:
  # *  loop forever {
  # *    update to new clock value;
  # *    evaluate (insertion of clock value);
  # *    evaluate (all tasks);
  # *    evaluate (deletion of clock value);
  # *  }
  def run_driver
    time = @clock.time(0)
    while 1
      synchronize do
        print("============================     EVALUATE SCHEDULE     =============================\n")
        evaluate(runtime.name, time.name, time, nil)
        print("============================     #{@tasks.length} TASKS     =============================\n")

        # Evaluate task queue
       # require 'ruby-debug'; debugger
        @tasks.each {|task| evaluate(task.program, task.name, task.insertions, task.deletions)}
        @tasks.clear # Clear task queue.
        evaluate(runtime.name, time.name, nil, time) # Clock delete current
        print("============================ ========================== =============================\n");

        # Check for new tasks or schedules, if none wait.
        while (@tasks.size == 0 && @schedule.cardinality == 0) do
          wait
        end
        print("============================     TASK ACQUIRED     =============================\n")
        if (@schedule.cardinality > 0)
          time = @clock.time(@schedule.min)
        else 
          time = @clock.time(@clock.current + 1)
        end
      end
    end
  end

  # * Helper function that calls the flusher and evaluator table functions.
  # * This function will evaluate the passed in tuples to fixedpoint (until
  # * no further tuples exist to evaluate).
  # * @param program The program that should be executed.
  # * @param name The table name to which the tuples refer.
  # * @param insertions The set of insertions to evaluate.
  # * @param deletions The set of deletions to evaluate.
  # * @throws UpdateException
  def evaluate(program, name, insertions, deletions)
#    require 'ruby-debug'; debugger
    insert = TupleSet.new("insert")
    delete = TupleSet.new("delete")
    insert << Tuple.new(@clock.current, program, name, insertions, deletions)
    # Evaluate until nothing remains.
    while (insert.size > 0 || delete.size > 0) do
      delta = nil
      while (insert.size > 0) do
        delta = @flusher.insert(insert, nil)
        delta = @evaluator.insert(delta, nil)
        insert.clear
        split(delta, insert, delete)
      end

      while(delete.size > 0) do
        delta = @flusher.insert(delete, nil)
        delta = @evaluator.insert(delta, nil)
        delete.clear
        split(delta, insert, delete)
      end
    end
  end

  def split(tuples, insertions, deletions)
    tuples.each do |tuple| 
      insert = tuple.clone;
      delete = tuple.clone;
      insert.value(Evaluator.Field::INSERTIONS, tuple.value(Evaluator.Field::INSERTIONS))
      insert.value(Evaluator.Field::DELETIONS, nil)
      delete.value(Evaluator.Field::INSERTIONS, nil)
      delete.value(Evaluator.Field::DELETIONS, tuple.value(Evaluator.Field::DELETIONS))

      insertions << insert
      deletions << delete
    end
  end
end
